src/Plugin.php

```diff
@@ -504,6 +505,9 @@ public function init() {
 		add_action('wp_enqueue_scripts', array(Cache::class, 'addWarmupAjaxToOutput'));
 		add_action('admin_enqueue_scripts', array(Cache::class, 'addWarmupAjaxToOutput'));
 
+		// Clear cached users on user update.
+		add_action( 'profile_update', array( UserRepository::class, 'clearUserCache' ) );
```

Comment:
Meiner Meinung nach an der Stelle vielleicht `set_user_role` hook nehmen. Eigentlich ändern ja nur Admin-Rollen die Rollen der User. Und jeder mögliche profile Update ist ja gar nicht interessant, sondern nur die Rollen.

https://developer.wordpress.org/reference/hooks/set_user_role/

`add_user_role` sollte nicht interessant sein, wir können doch nur eine user role zuweisen.

Reply:
Also meiner Meinung nach müsste auf jeden Fall z.B. noch der `updated_postmeta` Hook dazu, wenn sich User<->Owner Beziehungen ändern, muss der Cache auch invalidiert werden. Aber vielleicht können wir einen eignenen hook bauen für diese Beziehung, da das ja nicht die einzigen postmeta-Felder sind, die sich ändern.

Reply:
> Meiner Meinung nach an der Stelle vielleicht `set_user_role` hook nehmen. Eigentlich ändern ja nur Admin-Rollen die Rollen der User. Und jeder mögliche profile Update ist ja gar nicht interessant, sondern nur die Rollen.
> 
> https://developer.wordpress.org/reference/hooks/set_user_role/
> 
> `add_user_role` sollte nicht interessant sein, wir können doch nur eine user role zuweisen.

Naja, du kannst bei WordPress einem User mehrere Rollen geben. Theoretisch könnten wir auch einen Nutzenden dann noch zusätzlich CB_Manager machen. Dann würde, meinem Verständnis nach, die set_role nicht getriggert werden sondern nur die add_role.

Vielleicht ist der add_user_role Hook doch nicht so schlecht, weil der ja auch von der set_role gefeuert wird. Aber dann haben wir das Problem, dass bei einem entfernen der Rolle dieser Hook vielleicht nicht getriggert wird, das müsste ich nochmal ausprobieren.

Ich glaube, dass die Nutzenden davon auch nicht viel mitkriegen würden, da ja erst bei einem einloggen in das Backend der Cache wieder aufgebaut wird. Das würde dann vielleicht den Aufbau der ersten Seite für die Admins etwas langsamer machen aber mehr auch nicht.

Reply:
Habe jetzt erst gesehen, da ist es bereits schon drin.

Reply:
> Naja, du kannst bei WordPress einem User mehrere Rollen geben. Theoretisch könnten wir auch einen Nutzenden dann noch zusätzlich CB_Manager machen. Dann würde, meinem Verständnis nach, die set_role nicht getriggert werden sondern nur die add_role.

Stimmt.
 
> Vielleicht ist der add_user_role Hook doch nicht so schlecht, weil der ja auch von der set_role gefeuert wird. Aber dann haben wir das Problem, dass bei einem entfernen der Rolle dieser Hook vielleicht nicht getriggert wird, das müsste ich nochmal ausprobieren.

Ok dann wäre ich für `add_action` mit `add_user_role` und mit `remove_user_role`, anstatt dem `profile_update`. Dann haben wir die beiden Fälle abgedeckt.

> Ich glaube, dass die Nutzenden davon auch nicht viel mitkriegen würden, da ja erst bei einem einloggen in das Backend der Cache wieder aufgebaut wird. 

Mh ist das implizit im Cache konfiguriert? Dann habe ich den Cache noch nicht ganz verstanden.

> Das würde dann vielleicht den Aufbau der ersten Seite für die Admins etwas langsamer machen aber mehr auch nicht.

Ich habe den PR so verstanden, das er genau dafür ist.


Ich finde es wäre so sauberer, da wir nicht allgemeine Hooks verwenden sondern mit den speziellen Hooks auch dokumentieren was wir zu welchem Zweck invalidieren.

Reply:
> "An important distinction of wp_insert_post action is that it is fired after update_post_meta has been called." . Deshalb klinke ich mich mal hier vielleicht lieber in die bestehende savePostActions ein.

Oh das habe ich übesehen. Wieso wäre das ein Problem? Wir dürfen es nicht doppelt verwenden. Du wärst demnach eher für eine allgemeine anstatt eine spezifische Invalidierung des Caches? Also wie es aktuell schon implementiert ist in den savePostActions.

https://github.com/wielebenwir/commonsbooking/pull/1242/commits/dab777ae935d132d6ada3739c931371e7ea73bc1 Wie hier, würde ich zwei `add_action` für den add und remove hook erstellen. Der neue custom hook stände dann noch zur Debatte.

Reply:
@hansmorb Hat sich deine Meinung hier geändert oder drehe ich meine Änderungen wieder zurück?